doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,211 +1109,156 @@ Applying theorems
-
-   .. exn:: Unable to apply.
-
-      This happens if the conclusion of :token:`ident` does not match any of
-      the non-dependent premises of the type of :token:`term`.
-
-   .. tacv:: apply {+, @term} in {+, @ident}
-
-      This applies each :token:`term` in sequence in each hypothesis :token:`ident`.
-
-   .. tacv:: apply {+, @term with @bindings} in {+, @ident}
-
-      This does the same but uses the bindings to instantiate
-      parameters of :token:`term` (see :ref:`bindings`).
+.. _managingthelocalcontext:
 
-   .. tacv:: eapply {+, @term {? with @bindings } } in {+, @ident}
+Managing the local context
+------------------------------
 
-      This works as :tacn:`apply … in` but turns unresolved bindings into
-      existential variables, if any, instead of failing.
+.. tacn:: intro {? @ident } {? @where }
 
-   .. tacv:: apply {+, @term {? with @bindings } } in {+, @ident {? as @simple_intropattern}}
-      :name: apply … in … as
+   Applies the :tacn:`hnf` tactic until it finds an item that can be
+   introduced in the context by removing certain constructs in the goal.
+   If no item is found, the tactic fails.  The constructs have these forms:
+   (See examples :ref:`here <intro_examples>`.)
 
-      This works as :tacn:`apply … in` but applying an associated
-      :token:`simple_intropattern` to each hypothesis :token:`ident`
-      that comes with such clause.
+   `forall x : T, …`
+     `x` is a :term:`dependent premise`.  Removes `forall x`
+     from the goal and adds `x : T` to the context.
 
-   .. tacv:: simple apply @term in {+, @ident}
+   `A -> B`
+     `A` is a :term:`non-dependent premise`.  Removes `A ->` from
+     the goal and adds `H : A` to the context.  If `H` already exists, Coq uses
+     a fresh name.
 
-      This behaves like :tacn:`apply … in` but it reasons modulo conversion
-      only on subterms that contain no variables to instantiate and does not
-      traverse tuples. See :ref:`the corresponding example <simple_apply_ex>`.
+   `let x := c, …`
+     Removes `let x := c,` from the goal and adds `x := c : T`
+     to the context.
 
-   .. tacv:: {? simple} apply {+, @term {? with @bindings}} in {+, @ident {? as @simple_intropattern}}
-             {? simple} eapply {+, @term {? with @bindings}} in {+, @ident {? as @simple_intropattern}}
+   .. todo PR
+     Need an example of an existential variable
 
-      This summarizes the different syntactic variants of :n:`apply @term in {+, @ident}`
-      and :n:`eapply @term in {+, @ident}`.
+   We recommend always specifying :n:`@ident` so that the names of hypotheses don't
+   change as the proof is updated, making your proof easier to maintain.  For example,
+   if H exists in the context, Coq will consider using `H0`, `H1`, ... until it finds an
+   unused name.  Modifications to a proof can change automatically assigned names
+   that subsequent tactics likely refer to, making the proofs harder to maintain.
```

Comment:
This is a good recommendation, but there is no earlier mention that if provided, the `ident` will be used instead of an automatically generated name. We should add one.

Also, we could refer to the mangling name feature from here.